src/memory.rs

```diff
@@ -0,0 +1,153 @@
+//! Utilities for converting Kubernetes quantities to Java heap settings.
+
+use k8s_openapi::apimachinery::pkg::api::resource::Quantity;
+
+use crate::error::{Error, OperatorResult};
+use std::str::FromStr;
+
+#[derive(Clone, Debug, PartialEq, PartialOrd)]
+pub enum BinaryMultiple {
+    Kibi,
+    Mebi,
+    Gibi,
+    Tebi,
+    Pebi,
+    Exbi,
+}
+
+impl BinaryMultiple {
+    pub fn to_legacy(&self) -> String {
+        match self {
+            BinaryMultiple::Kibi => "k".to_string(),
+            BinaryMultiple::Mebi => "m".to_string(),
+            BinaryMultiple::Gibi => "g".to_string(),
+            BinaryMultiple::Tebi => "t".to_string(),
+            BinaryMultiple::Pebi => "p".to_string(),
+            BinaryMultiple::Exbi => "e".to_string(),
+        }
+    }
+
+    pub fn upscale(&self) -> Self {
+        match self {
+            BinaryMultiple::Kibi => BinaryMultiple::Kibi,
+            BinaryMultiple::Mebi => BinaryMultiple::Kibi,
+            BinaryMultiple::Gibi => BinaryMultiple::Mebi,
+            BinaryMultiple::Tebi => BinaryMultiple::Gibi,
+            BinaryMultiple::Pebi => BinaryMultiple::Tebi,
+            BinaryMultiple::Exbi => BinaryMultiple::Pebi,
+        }
+    }
+}
+
+impl FromStr for BinaryMultiple {
+    type Err = Error;
+
+    fn from_str(q: &str) -> OperatorResult<BinaryMultiple> {
+        let lq = q.to_lowercase();
+        match lq.as_str() {
+            "ki" | "kib" => Ok(BinaryMultiple::Kibi),
+            "mi" | "mib" => Ok(BinaryMultiple::Mebi),
+            "gi" | "gib" => Ok(BinaryMultiple::Gibi),
+            "ti" | "tib" => Ok(BinaryMultiple::Tebi),
+            "pi" | "pib" => Ok(BinaryMultiple::Pebi),
+            "ei" | "eib" => Ok(BinaryMultiple::Exbi),
+            _ => Err(Error::InvalidQuantityUnit {
+                value: q.to_string(),
+            }),
+        }
+    }
+}
+
+/// Easily transform K8S memory resources to Java heap options.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Memory {
+    value: f32,
+    unit: BinaryMultiple,
+}
+
+impl Memory {
+    /// Scale by the given factor. If the factor is less then one
+    /// the unit granularity is increased one level to ensure eventual
+    /// conversions to Java heap settings don't end up with zero values..
+    pub fn scale(&self, factor: f32) -> Self {
+        if factor < 1.0 && self.unit != BinaryMultiple::Kibi {
+            Memory {
+                value: self.value * factor * 1024.0,
+                unit: self.unit.upscale(),
+            }
+        } else {
+            Memory {
+                value: self.value * factor,
+                unit: self.unit.clone(),
+            }
+        }
+    }
+
+    /// The Java heap settings do not support fractional values therefore
+    /// this cannot be implemented without loss of precision.
+    pub fn to_java_heap(&self, factor: f32) -> String {
+        let scaled = self.scale(factor);
+        format!("-Xmx{:.0}{}", scaled.value, scaled.unit.to_legacy())
+    }
```

Comment:
1. Should the arbitrary scaling really be embedded into this call?
2. `Memory(2Gi).to_java_heap(0.8)` would be truncated down to 1Gi, which is a pretty far cry from the expected 1.6Gi. We should probably either hard-code it to output Mebibytes, or hard-code a certain number of significant digits rather than just ask whether it'd truncate down to zero.

Reply:
> It's a convenience since it's what we always have to do when computing the max Java heap.

`memory.to_java_heap(0.8)` doesn't save much typing compared to `(memory * 0.8).to_java_heap()`, and is (IMO) much less obvious about what's going on (especially without reading the definition of `to_java_heap` again).

> 2. It gets mapped to -Xmx1638m which assuming K8S does the right thing, is what we want.

Ok, then I'll need to read through this again because I don't understand where this is happening.

Reply:
Ah, you have the threshold where you try to avoid wasting more than 0.2x the base unit. I'd still prefer to always render to megabytes.. but meh.